From 2ae7a2397287a597e82c59a0e62658a494610c98 Mon Sep 17 00:00:00 2001
From: ZeromusXYZ <zeromus.xyz@gmail.com>
Date: Wed, 14 Aug 2024 09:54:14 +0200
Subject: [PATCH] [Game] Additional checks for trades (#1077)

- Added a check to the player trades that checks if the players actually have the amount of money they claim to put up for trade. Cancels the trade if either of them does not.
- Added additional warnings when trying to trade items that should not be able to be traded.
---
 AAEmu.Game/Core/Managers/TradeManager.cs | 65 +++++++++++++++---------
 1 file changed, 40 insertions(+), 25 deletions(-)

diff --git a/AAEmu.Game/Core/Managers/TradeManager.cs b/AAEmu.Game/Core/Managers/TradeManager.cs
index 643cad0eee..d39a29b48d 100644
--- a/AAEmu.Game/Core/Managers/TradeManager.cs
+++ b/AAEmu.Game/Core/Managers/TradeManager.cs
@@ -319,16 +319,24 @@ public void FinishTrade(Character owner, Character target, uint tradeId)
     {
         var tradeInfo = _trades[tradeId];
 
-        // TODO: check if this actually works correctly with the items-rewrite with the way packets are handled
-
-        /*
-         * TODO -
-         * try catch
-         * check if bound
-         * check untradable
-         */
+        // Validate Money (custom client protection)
+        if (tradeInfo.OwnerMoneyPutup > owner.Money)
+        {
+            CancelTrade(owner.ObjId, 0, tradeId); // Reason?
+            Logger.Error($"{owner.Name} ({owner.Id}) is putting up more money for trade than have {tradeInfo.OwnerMoneyPutup} > {owner.Money}, possible exploit or modified client!");
+            return;
+        }
+        if (tradeInfo.TargetMoneyPutup > target.Money)
+        {
+            CancelTrade(target.ObjId, 0, tradeId); // Reason?
+            Logger.Error($"{target.Name} ({target.Id}) is putting up more money for trade than have {tradeInfo.TargetMoneyPutup} > {target.Money}, possible exploit or modified client!");
+            return;
+        }
+
+        var hasErrors = 0;
         var tasksOwner = new List<ItemTask>();
         var tasksTarget = new List<ItemTask>();
+
         // Handle Money from Owner
         if (tradeInfo.OwnerMoneyPutup > 0)
         {
@@ -352,13 +360,15 @@ public void FinishTrade(Character owner, Character target, uint tradeId)
         {
             foreach (var item in tradeInfo.OwnerItems)
             {
-                target.Inventory.Bag.AddOrMoveExistingItem(ItemTaskType.Invalid, item);
-                //owner.Inventory.RemoveItem(item, false);
-                tasksOwner.Add(new ItemRemove(item));
-                //item.Slot = -1;
-                //var newItem = target.Inventory.AddItem(item);
-                //tasksTarget.Add(new ItemAdd(newItem));
-                tasksTarget.Add(new ItemAdd(item));
+                if (target.Inventory.Bag.AddOrMoveExistingItem(ItemTaskType.Invalid, item))
+                {
+                    tasksOwner.Add(new ItemRemove(item));
+                    tasksTarget.Add(new ItemAdd(item));
+                }
+                else
+                {
+                    hasErrors++;
+                }
             }
         }
         // Handle Items from Target
@@ -366,21 +376,26 @@ public void FinishTrade(Character owner, Character target, uint tradeId)
         {
             foreach (var item in tradeInfo.TargetItems)
             {
-                owner.Inventory.Bag.AddOrMoveExistingItem(ItemTaskType.Invalid, item);
-                //target.Inventory.RemoveItem(item, false);
-                tasksTarget.Add(new ItemRemove(item));
-                //item.Slot = -1;
-                //var newItem = owner.Inventory.AddItem(item);
-                // tasksOwner.Add(new ItemAdd(newItem));
-                tasksOwner.Add(new ItemAdd(item));
+                if (owner.Inventory.Bag.AddOrMoveExistingItem(ItemTaskType.Invalid, item))
+                {
+                    tasksTarget.Add(new ItemRemove(item));
+                    tasksOwner.Add(new ItemAdd(item));
+                }
+                else
+                {
+                    hasErrors++;
+                }
             }
         }
 
+        // Trade complete, remove ID and send item task packets
         _trades.Remove(tradeId);
         owner.SendPacket(new SCTradeMadePacket(ItemTaskType.Trade, tasksOwner, new List<ulong>()));
         target.SendPacket(new SCTradeMadePacket(ItemTaskType.Trade, tasksTarget, new List<ulong>()));
-        Logger.Info($"Trade Id:{tradeId} finished." +
-                    $" Owner Items/Money: {tradeInfo.OwnerItems.Count}/{tradeInfo.OwnerMoneyPutup}." +
-                    $" Target Items/Money: {tradeInfo.TargetItems.Count}/{tradeInfo.TargetMoneyPutup}");
+        Logger.Info($"Trade Id:{tradeId} finished. Owner {owner.Name} ({owner.Id}) Items/Money: {tradeInfo.OwnerItems.Count}/{tradeInfo.OwnerMoneyPutup} <=> Target {target.Name} ({target.Id}) Items/Money: {tradeInfo.TargetItems.Count}/{tradeInfo.TargetMoneyPutup}");
+        if (hasErrors > 0)
+        {
+            Logger.Error($"{hasErrors}item(s) could not be trade for tradeId: {tradeId} between {owner.Name} ({owner.Id}) and {target.Name} ({target.Id}), possible exploit or modified client!");
+        }
     }
 }
